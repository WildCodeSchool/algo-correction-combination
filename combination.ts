export const combine = (char: string, chars: string[], length: number): string[] => {
    let result = [char]
    for (let i = 0; i < chars.length; i++) {
        let code = char + chars[i]    
        if (length > 2) {
            result = result.concat(combine(code, chars, length-1))
        } else {
            result.push(code)
        }
    }
    return result
}
const combination = (chars: string[], length: number):string[] => {
    return chars.flatMap(char => combine(char, chars, length)).sort()
}

export default combination
