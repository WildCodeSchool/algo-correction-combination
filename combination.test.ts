import combination from "./combination"

describe('combination', () => {
    it('combinate all chars', () => {
        expect(combination(['a','b','c'], 2)).toEqual(['a','b','c','aa','ab','ac','ba','bb','bc','ca','cb','cc'])
    })
    it('combinate all chars', () => {
        expect(combination(['a','b','c'], 3)).toMatchSnapshot()
    })
})