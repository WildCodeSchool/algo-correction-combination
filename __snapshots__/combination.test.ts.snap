// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`combination combinate all chars 1`] = `
Array [
  "a",
  "aa",
  "aaa",
  "aab",
  "aac",
  "ab",
  "aba",
  "abb",
  "abc",
  "ac",
  "aca",
  "acb",
  "acc",
  "b",
  "ba",
  "baa",
  "bab",
  "bac",
  "bb",
  "bba",
  "bbb",
  "bbc",
  "bc",
  "bca",
  "bcb",
  "bcc",
  "c",
  "ca",
  "caa",
  "cab",
  "cac",
  "cb",
  "cba",
  "cbb",
  "cbc",
  "cc",
  "cca",
  "ccb",
  "ccc",
]
`;

exports[`combine combine one char with chars 1`] = `
Array [
  "a",
  "aa",
  "aaa",
  "aab",
  "aac",
  "ab",
  "aba",
  "abb",
  "abc",
  "ac",
  "aca",
  "acb",
  "acc",
]
`;
